### DISCUSSION TITLE
Custom tilemap and shader help

### DISCUSSION THREAD

**Discussion by IsaacGaches** [bevyengine/bevy#9681] [Q&A]
I want to make a game like terraria or starbound so I need to render the tile based world and I want to do it in the most efficient way I reasonably can (a custom tilemap). Please do not recommend tilemap crates because I want to do it from scratch for a list of reasons I wont get into. Also the built in 2d mesh has a number of issues so I'm not using that.

I just need directions on how to at least make a wgsl shader appear on the screen with bevy (2d) so I can make a custom tilemap.

Sorry this isn't very formal or descriptive I'm tired...

**Comment by JMS55** [comment]
If you don't want to use a tilemap crate, then you should look at their source code and see how they're implemented.

**Comment by cwalter5** [comment]
I was looking into the same thing and decided my best option would be to write my own shader as the existing crates wouldn’t be performant enough. Bevy fast tilemap did come close but unfortunately it copies the whole tilemap every frame which with a world of that scale kills pretty much any integrated gpu.

I was suggested to look into https://github.com/bevyengine/bevy/blob/main/examples/shader/post_processing.rs but sadly haven’t had the time to look into this further. Hope that helps.

**Reply by Droggelbecher** [reply]
Hey @cwalter5 author of bevy fast tilemap here. It only copies the whole map in frames in which you make changes, but I agree that that is less than ideal. I'm still looking for ways to improve this, if you have found a good solution I'd be very happy to hear about it :)

**Reply by Droggelbecher** [reply]
Oh btw, just for perspective: the "updates" example which updates a 1024x1024 tilemap every single frame runs with >200fps on my laptop (RTX 3070 Laptop GPU), so this should be quite usable as-is for a relatively wide array of projects.
